### DISCUSSION TITLE
Displaying Form Data with Dynamic Field Types Using Filament's Form Repeater

### DISCUSSION THREAD

**Discussion by adiartawibawa** [filamentphp/filament#12520] [Help]
### Package

Form builder

### Package Version

v3.1

### How can we help you?

I am currently working on a web-based form application using Laravel 10 and Filament PHP. I have a form table with columns containing data as follows:

```json
[
    {
        "value": "John Doe",
        "desc": "Full Name",
        "type": "string",
        "valid": false
    },
    {
        "value": "Elm Street No. 13",
        "desc": "Address",
        "type": "text",
        "valid": false
    },
    {
        "value": "0142365789",
        "desc": "Phone Number",
        "type": "string",
        "valid": false
    }
]
```

Now, I want to display this data using the Form Repeater provided by Filament. Below is a snippet of my code:

```php
Step::make('Form')
    ->description('Form submission')
    ->schema([
        Forms\Components\Repeater::make('form')->schema(function (Get $get): array {
            switch ($get('type')) {
                case 'string':
                    $formField = TextInput::make('value');
                    break;
                case 'text':
                    $formField = Textarea::make('value');
                    break;
                case 'image':
                    $formField = SpatieMediaLibraryFileUpload::make('value')->getUploadedFiles();
                    break;
                case 'file':
                    $formField = FileUpload::make('value')->downloadable();
                    break;
                case 'date':
                    $formField = DatePicker::make('value')->displayFormat(function (): string {
                        return 'd/m/Y';
                    });
                    break;
                default:
                    $formField = TextInput::make('value');
            }
            return [
                $formField,
                Textarea::make('note'),
                Toggle::make('valid'),
            ];
        })->addable(false)
            ->deletable(false)
            ->reorderable(false)
            ->reorderableWithDragAndDrop(false)
            ->columnSpanFull(),
    ])
```

When I tried running the code I didn't encounter any errors at all and the data appeared by what was stored in the table until finally I realized that the data did not appear according to the type of each data. For example, if the data type is a string, a `TextInput` should be displayed, and if it's a file, a `FileInput` should be displayed. The problem is, that I'm not sure how to correctly access the data type from the form column before displaying it. When I tried adding `dd($get('type))`, it returned `null`.

Can anyone help me solve this issue? Any advice or answers would be greatly appreciated. Thank you.

**Comment by leandrocfe** [comment]
You should assign a unique identifier to each of your fields instead of using `make('value')` for all of them.

**Reply by adiartawibawa** [reply]
Thank you for the advice. But I'm still curious about how I can access the type of data stored in the table

**Comment by ismaeltj9** [comment]
> ### Package
> Form builder
> 
> ### Package Version
> v3.1
> 
> ### How can we help you?
> I am currently working on a web-based form application using Laravel 10 and Filament PHP. I have a form table with columns containing data as follows:
> 
> ```json
> [
>     {
>         "value": "John Doe",
>         "desc": "Full Name",
>         "type": "string",
>         "valid": false
>     },
>     {
>         "value": "Elm Street No. 13",
>         "desc": "Address",
>         "type": "text",
>         "valid": false
>     },
>     {
>         "value": "0142365789",
>         "desc": "Phone Number",
>         "type": "string",
>         "valid": false
>     }
> ]
> ```
> 
> Now, I want to display this data using the Form Repeater provided by Filament. Below is a snippet of my code:
> 
> ```
> Step::make('Form')
>     ->description('Form submission')
>     ->schema([
>         Forms\Components\Repeater::make('form')->schema(function (Get $get): array {
>             switch ($get('type')) {
>                 case 'string':
>                     $formField = TextInput::make('value');
>                     break;
>                 case 'text':
>                     $formField = Textarea::make('value');
>                     break;
>                 case 'image':
>                     $formField = SpatieMediaLibraryFileUpload::make('value')->getUploadedFiles();
>                     break;
>                 case 'file':
>                     $formField = FileUpload::make('value')->downloadable();
>                     break;
>                 case 'date':
>                     $formField = DatePicker::make('value')->displayFormat(function (): string {
>                         return 'd/m/Y';
>                     });
>                     break;
>                 default:
>                     $formField = TextInput::make('value');
>             }
>             return [
>                 $formField,
>                 Textarea::make('note'),
>                 Toggle::make('valid'),
>             ];
>         })->addable(false)
>             ->deletable(false)
>             ->reorderable(false)
>             ->reorderableWithDragAndDrop(false)
>             ->columnSpanFull(),
>     ])
> ```
> 
> When I tried running the code I didn't encounter any errors at all and the data appeared by what was stored in the table until finally I realized that the data did not appear according to the type of each data. For example, if the data type is a string, a `TextInput` should be displayed, and if it's a file, a `FileInput` should be displayed. The problem is, that I'm not sure how to correctly access the data type from the form column before displaying it. When I tried adding `dd($get('type))`, it returned `null`.
> 
> Can anyone help me solve this issue? Any advice or answers would be greatly appreciated. Thank you.

I need exactly the same, did you have some solution?
